File: Valid Parentheses.py
# Leetcode Problem 20: Valid Parentheses
# Python CODE
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        mp = {
            ')': '(',
            '}': '{',
            ']': '['
        }

        for ch in s:
            if ch in mp:
                if not stack or stack[-1] != mp[ch]:
                    return False
                stack.pop()
            else:
                stack.append(ch)

        return len(stack) == 0

File: Valid Parentheses.cpp
// Leetcode Problem 20: Valid Parentheses
// C++ CODE
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;

        for (char ch : s) {
            if (ch == '(' || ch == '{' || ch == '[') {
                st.push(ch);
            } else {
                if (st.empty())
                    return false;

                if ((ch == ')' && st.top() != '(') ||
                    (ch == '}' && st.top() != '{') ||
                    (ch == ']' && st.top() != '['))
                    return false;

                st.pop();
            }
        }

        return st.empty();
    }
};